### ISSUE TITLE
TextInputConnection should let objects listen for IME connection closure

### ISSUE THREAD

**Issue by matthew-carroll** [flutter/flutter#117771]
An IME connection is established with thee pieces: `TextInput`, a `TextInputConnection`, and a `TextInputClient`.

```dart
final myConnection = TextInput.attach(myInputClient, myConfig);
```

`TextInputConnection` has a `bool` that reports `isAttached`, but there's no way to listen for a change to this attachment.

The `TextInputClient` has a method called `connectionClosed`, which, I assume, represents a notification that the IME connection has closed. However, policies about when to open and close a connection are usually independent from the policies that apply text input messages from the platform. Therefore, the use of `TextInputConnection` should be able to exist independently from the code that implements `TextInputClient`, and this requires that `TextInputConnection` receive a notification when its connection is closed.

```
Flutter 3.7.0-1.2.pre • channel beta • https://github.com/flutter/flutter.git
Framework • revision c29b09b878 (9 days ago) • 2022-12-19 18:30:12 -0600
Engine • revision 8c99b0feaf
Tools • Dart 2.19.0 (build 2.19.0-444.2.beta) • DevTools 2.20.0
```

**Comment by matthew-carroll** [comment]
CC @justinmc @Renzo-Olivares

**Comment by justinmc** [comment]
This concern seems valid to me.

Looking at the current code, it does seem a little convoluted to call connectionClosed on TextInputConnection's TextInputClient instance and then immediately call back to the TextInputConnection.

Would it work for your use case if it went TextInput -> TextInputConnection -> TextInputClient instead?  That way TextInput would inform TextInputConnection that it's being closed without needing to rely on the implementation of TextInputClient's connectionClosed method.

### How it works now

 1. TextInput receives the platform channel message `TextInputClient.onConnectionClosed` and calls `TextInputClient.connectionClosed`.
https://github.com/flutter/flutter/blob/b7881e5b64e28a84a66ab30b29d35484428ea062/packages/flutter/lib/src/services/text_input.dart#L1899-L1900

 2. If using EditableText, then `TextInputConnection.connectionCloseReceived` is called.
https://github.com/flutter/flutter/blob/b7881e5b64e28a84a66ab30b29d35484428ea062/packages/flutter/lib/src/widgets/editable_text.dart#L3095-L3097

### Proposed changes

https://github.com/flutter/flutter/pull/118283

**Comment by matthew-carroll** [comment]
> Would it work for your use case if it went TextInput -> TextInputConnection -> TextInputClient instead

Yes, I think so. The most important part is that whatever object holds the `TextInputConnection` shouldn't need to also hold the `TextInputClient` and vis-a-versa. The job of the connection is independent from the job of applying deltas, and both of those jobs might want to know when the connection closes.

**Comment by jzhgithub18** [comment]
can i get a way to to set the Candidate Window postion. when i input the Chinese language. 
int windows and macos platform
final myConnection = TextInput.attach(myInputClient, myConfig);

**Comment by justinmc** [comment]
@jzhgithub18 Can you file a separate issue for that?

By Candidate Window you mean the composing input window? (screenshot is not Flutter)

![Screenshot from 2023-04-13 10-32-52](https://user-images.githubusercontent.com/389558/231839041-9700014d-90d1-4a41-917c-d49390a38803.png)

**Comment by jzhgithub18** [comment]
@justinmc Thank you for your reply
yes, It always appears in the upper left corner of the window when I am using it.
![image](https://user-images.githubusercontent.com/112080579/231928256-ea222909-97ce-4d8c-9ab4-b808c23d4700.png)

I expectation the Candidate Window can be normal in the below of Pinyin.
![image](https://user-images.githubusercontent.com/112080579/231928443-542e2095-01a1-42d5-b16b-a8d9e55aea39.png)

and i know the textfiled is ok, but i don't konw how to do it.

my flutter version is Channel stable, 3.7.7, on Microsoft Windows

**Comment by agata** [comment]
I am experiencing the same issue and am facing difficulties. Like @jzhgithub18, the Japanese IME appears in the top left corner or in an incorrect position.

When using a TextField, the IME appears directly below the caret, which is as expected (I’ve attached a video showing the IME behavior with a TextField. This is the typical behavior for macOS applications).

I have confirmed this issue with the following version and platform:

Flutter 3.22.2
macOS 14.5

https://github.com/user-attachments/assets/681eb28f-0309-4a69-9e11-bc6f88e47492
